ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -0,0 +1,502 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Abstract class to handle replication for given partitions. Responsible for the following
+ * 1. Initialize class members.
+ * 2. Create replica threads and distribute partitions amongst the threads
+ * 3. Set up replica token persistor used to recover from shutdown/crash
+ * 4. Initialize and shutdown all the components required to perform replication
+ */
+public abstract class ReplicationEngine {
+
+  private final ReplicationConfig replicationConfig;
+  private final ClusterMap clusterMap;
+  private final ScheduledExecutorService scheduler;
+  private final AtomicInteger correlationIdGenerator;
+  private final ConnectionPool connectionPool;
+  private final NotificationSystem notification;
+  private final Map<String, DataNodeRemoteReplicaInfos> dataNodeRemoteReplicaInfosPerDC;
+  private final StoreKeyFactory storeKeyFactory;
+  private final List<String> sslEnabledDatacenters;
+  private final Map<String, List<ReplicaThread>> replicaThreadPools;
+  private final StoreKeyConverterFactory storeKeyConverterFactory;
+  private final String transformerClassName;
+
+  protected final DataNodeId dataNodeId;
+  protected final MetricRegistry metricRegistry;
+  protected final ReplicationMetrics replicationMetrics;
+  protected final FindTokenFactory factory;
+  protected final Logger logger = LoggerFactory.getLogger(getClass());
+  protected final Map<PartitionId, PartitionInfo> partitionsToReplicate;
+  protected final Map<String, List<PartitionInfo>> partitionGroupedByMountPath;
+  protected final Map<String, Integer> numberOfReplicaThreads;
+  protected ReplicaTokenPersistor persistor = null;
+
+  protected static final short Replication_Delay_Multiplier = 5;
+  protected static final String replicaTokenFileName = "replicaTokens";
+  private static final short Crc_Size = 8;
+
+  public ReplicationEngine(ReplicationConfig replicationConfig, ClusterMapConfig clusterMapConfig,
+      StoreKeyFactory storeKeyFactory, ClusterMap clusterMap, ScheduledExecutorService scheduler, DataNodeId dataNode,
+      ConnectionPool connectionPool, MetricRegistry metricRegistry, NotificationSystem requestNotification,
+      StoreKeyConverterFactory storeKeyConverterFactory, String transformerClassName) throws ReplicationException {
+    this.replicationConfig = replicationConfig;
+    this.storeKeyFactory = storeKeyFactory;
+    try {
+      this.factory = Utils.getObj(replicationConfig.replicationTokenFactory, storeKeyFactory);
+    } catch (ReflectiveOperationException e) {
+      logger.error("Error on getting replicationTokenFactory", e);
+      throw new ReplicationException("Error on getting replicationTokenFactory");
+    }
+    this.replicaThreadPools = new HashMap<>();
+    this.replicationMetrics = new ReplicationMetrics(metricRegistry, clusterMap.getReplicaIds(dataNode));
```

Comment:
Minor: do you think it's worthwhile to make `ReplicationManager` and `CloudBackupManager` use separate metrics? (currently, they are sharing `replicationMetrics`)

Reply:
That's a good point. Will address it in backupManager PR.